### DISCUSSION TITLE
GCP PubSub Source Flow Control

### DISCUSSION THREAD

**Discussion by atibdialpad** [vectordotdev/vector#12990] [Q&A] [answered]
Hey @bruceg, 
Thanks for building the gcp pubsub source. Appreciate !!! 

I have a question regarding the flow control of messages pulled in by this PubSub Source.
I am running a GKE Vector Deployment which pulls in messages from a GCP PubSub subscription. As part of the Deployment I am running a bunch of pods each with 2.5-3 vCPUs request-limits. But I see the Vector Pods are only using 0.05-0.07 vCPU on an average however there is significant and constant backlog of messages in the GCP Subscription.

I am trying to figure out the reason for this backlog and could come up with 1 yet to be tested hypothesis : The subscription messages are itself being pulled not fast enough causing the backlog to build in the subscription. Vector receives messages at a very low rate thereby not requiring it to vertically scale up the CPU needs to what is being requested. 

To it, my question : is there a way to identify where the bottleneck is via vector internal metrics or some debugging magic ? Any pro-tip would be super helpful.
 
@jszwedko

**Comment by jszwedko** [comment]
Hi @atibdialpad !

I'd suggest taking a look at the `utilization` metric for your components to determine if there is a bottleneck in Vector's processing that is causing backpressure or if it is the source that is not feeding fast enough. A value lower than `1` indicates that the component is spending some time idle so the last one that is `1` would indicate it is a bottleneck. If all of the components fed by the GCP PubSub source are idle, that could indicate the source could be receiving faster, but is not.

cc/ @bruceg for any other thoughts since he worked on that source.

**Comment by bruceg** [comment]
The PubSub source uses a gRPC streaming client to receive the data from GCP, which should result minimize the bottlenecks on the Vector side. I would look at the `utilization` metric as Jesse suggested. You could also try replacing the output sink to something known to have minimal overhead (ie `file` or even `blackhole`) and see if that moves the needle.

**Comment by atibdialpad** [comment]
Thanks folks for the suggestions. I will try them out today and report back.

**Comment by atibdialpad** [comment]
I observed the utilisation of all the stages/transforms in the pipeline. Below is a timeseries of the utilisation metric from the very first vector transform (just after the pubsub source) from each of the pod in the Deployment. As can be seen the transform is hugely under utilised. The metric is in order of 0.003-0.004 (0.3% to 0.4%)
<img width="1616" alt="Screenshot 2022-06-07 at 8 49 54 PM" src="https://user-images.githubusercontent.com/48474281/172418655-5b865175-d9f0-4e58-8504-b13a140f5485.png">

One more observation: There was 4 hr window when the number of messages in the topic ramped up and caused a huge increase in the number of unack'ed msgs in the subscription. The Vector utilisation of the same (first) transform increased to 0.25-0.30 (25%-30%) (though not much increase in the POD's CPU Utilisation). The utilisation went to earlier/normal low levels when the backlog went back to the normal level (~30K-40K unack'ed msgs)
<img width="1082" alt="Screenshot 2022-06-07 at 8 45 00 PM" src="https://user-images.githubusercontent.com/48474281/172419574-f3054109-0da4-4193-8f51-5bb63bccbe83.png">

So it does look to me that the gcppub subscription <-> vector pubsub src is the bottleneck.

My Config:
```
[sources.cloud_log_source]
  type = "gcp_pubsub"
  ack_deadline_seconds = 10                                                                                                                                                                           
  project = "fststaging"
  subscription = "log-ingest-vector"
  credentials_path = "gcp_ingestion_key_staging.json"
```

**Comment by atibdialpad** [comment]
new observation, I see a ton of error logs in the google_pubsub src trying to pull messages from the subscription :  
<img width="1895" alt="Screenshot 2022-06-07 at 10 53 32 PM" src="https://user-images.githubusercontent.com/48474281/172444983-d1246c49-7cd0-4024-978c-3199de61025b.png">

Exact Error :  
```
severity: "ERROR"
textPayload: "2022-06-07T14:07:56.633928Z ERROR source{component_kind="source" component_id=cloud_log_source component_type=gcp_pubsub component_name=cloud_log_source}: vector::internal_events::gcp_pubsub: Failed to fetch events. error=status: InvalidArgument, message: "Request contains an invalid argument.", details: [], metadata: MetadataMap { headers: {"content-disposition": "attachment"} } error_code="failed_fetching_events" error_type="request_failed" stage="receiving""

severity: "ERROR"
textPayload: "2022-06-07T14:07:56.634022Z  INFO source{component_kind="source" component_id=cloud_log_source component_type=gcp_pubsub component_name=cloud_log_source}: vector::sources::gcp_pubsub: Retrying after timeout. timeout_secs=1.0"
```
which is basically this https://github.com/vectordotdev/vector/blob/a4ff8b36e166363a2ffc0662de59d93e1f780c16/src/internal_events/gcp_pubsub.rs#L59 error 
This could be the errors why the subscription <-> vector path is bottlenecked, right ? @bruceg @jszwedko 
Need to check the error further

**Comment by atibdialpad** [comment]
I might be hitting this https://github.com/vectordotdev/vector/issues/12660 . The error and throughput observation seems identical. I will try with the latest image tomorrow and report.

**Comment by atibdialpad** [comment]
[Update]
With the latest image, the "invalid argument" errors are not seen anymore.

I see this error 3-4 times in an hour (far less frequent then the earlier invalid arg error). Doesn't look much harmful.
```
textPayload: "2022-06-08T11:30:50.941488Z ERROR source{component_kind="source" component_id=cloud_log_source component_type=gcp_pubsub component_name=cloud_log_source}: vector::internal_events::gcp_pubsub: Failed to fetch events. error=status: Unavailable, message: "The service is currently unavailable.", details: [], metadata: MetadataMap { headers: {"content-disposition": "attachment"} } error_code="failed_fetching_events" error_type="request_failed" stage="receiving""

severity: "ERROR"
textPayload: "2022-06-08T11:30:50.941704Z  INFO source{component_kind="source" component_id=cloud_log_source component_type=gcp_pubsub component_name=cloud_log_source}: vector::sources::gcp_pubsub: Retrying after timeout. timeout_secs=0.5"
timestamp: "2022-06-08T11:30:50.941879019Z"
```

However the backlog and underutilisation still persists.

**Comment by atibdialpad** [comment]
I did a bit of deep dive into this issue and also compared the performance with a logstash GKE Deployment (same cluster, same Node types, same request and resources). We found out that the logstash [gcp_pubsub_input](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-google_pubsub.html) plugin performs just fine and keeps the number of backlogged msg well under limits.

Looking at the GCP PubSub Subscription Monitoring [Metrics](https://cloud.google.com/monitoring/api/metrics_gcp#gcp-pubsub) it appears that the poor performance in this setup has to do with Acknowledgement Request Latencies (subscription/ack_latencies ) **~3-5 mins for Vector but ~60 ms for Logstash**
<img width="1489" alt="Screenshot 2022-06-10 at 1 54 16 PM" src="https://user-images.githubusercontent.com/48474281/173024055-adf4c337-8b3b-4ad5-8267-268a2fddfab0.png">

@bruceg  Could the fact that Vector is using streaming_pull while logstash is using just Normal pull play a role here ? (Need to see if streaming pull involves ack'ing in bulk )

The subscription/ack_message_count is almost similar 
<img width="1489" alt="Screenshot 2022-06-10 at 2 03 08 PM" src="https://user-images.githubusercontent.com/48474281/173026016-646efcbe-a194-4440-837a-b8e2ce284c9c.png">


but when I look at how many ack requests timed out (subscription/expired_ack_deadlines_count) Almost 50/sec ack requests timed out for Vector but ~0 for Logstash
<img width="1489" alt="Screenshot 2022-06-10 at 2 24 45 PM" src="https://user-images.githubusercontent.com/48474281/173029888-80687b82-aab9-490f-8bc0-001bd7eba526.png">
which looks like it is because of the ack latency.


Further, just looking at the number of ack requests (subscription/pull_ack_request_count) The Vector number is very high **~200K /s** while logstash is **~10 /s**
<img width="1489" alt="Screenshot 2022-06-10 at 2 14 54 PM" src="https://user-images.githubusercontent.com/48474281/173028127-a3bc2b94-8073-444b-b53c-71a0e4c1f3fb.png">
which I feel is because of the retries since many Vector ack times out (as seen in the earlier pic)

So, the root cause atm looks like the high ack latency. Will have to check if we can play around with https://vector.dev/docs/reference/configuration/sources/gcp_pubsub/#ack_deadline_seconds ? @jszwedko @bruceg any suggestions here ?

Another question ?
Are we not using [FlowControl](https://cloud.google.com/pubsub/docs/pull#flow_control) for the streaming pull request ? Will Adding an explicity max_outstading_message/bytes help ? Ref [gcp pub sub api](https://cloud.google.com/pubsub/docs/reference/rpc/google.pubsub.v1#streamingpullrequest), [proto](https://github.com/vectordotdev/vector/blob/a89a8cee5679613dea5582d2f44fcda2a82c3425/proto/google/pubsub/v1/pubsub.proto#L186), not using flow control in [vector code](https://github.com/vectordotdev/vector/blob/81658ec39509c08717967b8b489d2bfe49bb65af/src/sources/gcp_pubsub.rs#L342)
Vs
Logstash seems to be using it [cofig](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-google_pubsub.html#plugins-inputs-google_pubsub-max_messages) : [code](https://github.com/logstash-plugins/logstash-input-google_pubsub/blob/6c81bae14088052af0b8a07de4e24b0f5dbbeb47/lib/logstash/inputs/google_pubsub.rb#L264)

**Reply by bruceg** [reply]
> @bruceg Could the fact that Vector is using streaming_pull while logstash is using just Normal pull play a role here ? (Need to see if streaming pull involves ack'ing in bulk )

It could be. After today's [stream request fix](https://github.com/vectordotdev/vector/pull/13203) we don't seem to be limited by our local processing speed, however. If there are still speed differences, it may have more to do with GCP limitations on stream bandwidth than Vector. Regarding acking in bulk, the streaming support allows us to stream back arrays of acknowledgements as fast as they happen.

**Comment by atibdialpad** [comment]
@bruceg Ping !! :-) 
Also, JFYI: I tried by setting a 10 min ack deadline on both the Google PubSub Subscription as well as in the Vector PubSub Source config but no luck.

**Reply by jszwedko** [reply]
Thanks for the investigation @atibdialpad ! @bruceg will be looking into this soon once he gets to a good context switching point.

Could you share the `gcp_pubsub` source config you are using?

**Comment by atibdialpad** [comment]
Thanks @jszwedko 
```
[sources.cloud_log_source]
  type = "gcp_pubsub"
  ack_deadline_seconds = 600 # Tried with 10 seconds as well
  retry_delay_seconds = 1 # Tried with 0.5s
  project = "fststaging"
  subscription = "log-ingest-vector-v3"
  # credentials_path = "gcp_test_key.json" # a) Tried with the json key b) Tried without Json key with GKE Workload Identify
  acknowledgements.enabled = false # All sinks have this as false for experimentation.
```

**Comment by bruceg** [comment, marked as answer]
We just [merged a PR](https://github.com/vectordotdev/vector/pull/13203) that should at least improve the handling of this problem. Could you try out the next nightly after tonight to see if it improves the situation for you?

**Comment by atibdialpad** [comment]
Sure @bruceg will try the latest nightly. Thanks for looking into this.

**Comment by atibdialpad** [comment]
I have been running the Vector Deployment since the last 2 days and the fix look super solid !!!!! Thanks @bruceg once again.

**Reply by bruceg** [reply]
Great, thanks for confirming.
